### PR TITLE
Add gas limit to `getBalance` and `getAllowance`

### PR DESCRIPTION
**Pull request by abandeali1**
- Only forwards 4999 gas to `token.balanceOf` and `token.allowance` in order to prevent state changes when calling malicious tokens
- Addresses issue #107

### CODE REVIEW

File: contracts/Exchange.sol [hunk 1]
```diff
@@ -31,6 +31,7 @@ contract Exchange is SafeMath {
     uint8 constant ERROR_ROUNDING_ERROR_TOO_LARGE = 2;          // Rounding error too large
     uint8 constant ERROR_INSUFFICIENT_BALANCE_OR_ALLOWANCE = 3; // Insufficient balance or allowance for token transfer
 
+    uint16 constant CONSTANT_EXTERNAL_CALL_GAS_LIMIT = 4999;    // Gas limit at which state update cannot occur
```
Comment (willwarren89) [inline contracts/Exchange.sol]: Perhaps rename to EXTERNAL_QUERY_GAS_LIMIT.

File: test/ts/exchange/core.ts [hunk 2]
```diff
@@ -517,6 +518,22 @@ contract('Exchange', (accounts: string[]) => {
       assert.deepEqual(newBalances, balances);
     });
 
+    it('should throw if getBalance or getAllowance attempts to change state and shouldThrowOnInsufficientBalanceOrAllowance = false', async () => {
+      const maliciousToken = await MaliciousToken.new();
+      await maliciousToken.approve(Proxy.address, INIT_ALLOW, { from: taker });
```
Comment (fabioberger) [inline test/ts/exchange/core.ts]: Rename `INIT_ALLOW ` to `INITIAL_ALLOWANCE`

File: contracts/tokens/MaliciousToken.sol [hunk 3]
```diff
@@ -0,0 +1,29 @@
+pragma solidity 0.4.11;
+
+import "./../base/StandardToken.sol";
+
+contract MaliciousToken is StandardToken {
```
Comment (ethers) [inline contracts/tokens/MaliciousToken.sol]: not sure about Truffle but could this contract be put in a directory like `test/contracts/MaliciousToken.sol`?  (less preferable is `contracts/test/MaliciousToken.sol`)

File: contracts/Exchange.sol (hunk 1 above)
Comment (maurelian) [inline contracts/Exchange.sol]: I'm slightly in favor of `EXTERNAL_QUERY_GAS_LIMIT`, as it seems more indicative of the reason for it's existence.

As a comment, I'd find `//  changes to state require at least 5000 gas` slightly more direct.

Comment (abandeali1) [inline contracts/Exchange.sol]: Agreed, I am generally terrible at naming variables.

File: contracts/tokens/MaliciousToken.sol (hunk 3 above)
Comment (abandeali1) [inline contracts/tokens/MaliciousToken.sol]: `test/contracts` didn't work, but `contracts/test` is still better than the current structure.

**Pull request merged**